civihr_employee_portal/src/View/ViewsResultDiff.php

```diff
@@ -0,0 +1,284 @@
+      }
+    }
+    $i--;
+  }
+
+  /**
+   * To compare data generated between 2 different moments
+   *
+   * @return ViewsResultDiff
+   */
+  public function compareData() {
+    $this->compareDataRecursive(
+      $this->oldData,
+      $this->newData,
+      $this->changes
+    );
+    return $this;
+  }
+
+  /**
+   * To retrieve the current data of the views configured to diff
+   * this function will be called by storeNewData or storeOldData
+   * to be able to store the data produced by the views for later
+   * comparison
+   *
+   * @return array
+   *     The data retrieved from one of more views,
+   *     joined on specific fields for later comparison
+   */
+  private function getCurrentData($settings = null) {
+    $data = [];
+    $settings = !$settings ? $this->viewsSettings : $settings;
+
+    foreach ($settings as $setting) {
+      // if it is a group then apply recursion
+      if (isset($setting['group'])) {
+        $data[] = ['group' => $this->getCurrentData($setting['group'])];
+      } else if (isset($setting['view'])) {
+        // if it is a view it must be previewed to get the results
+        $view = views_get_view($setting['view']);
+
+        // Setting Exposed Filters
+        // backing up $_GET
+        $_GETbkp = $_GET;
+        // add exposed filter keys to $_GET
+        $_GET = isset($setting['exposedFilters']) ? $setting['exposedFilters'] + $_GET : $_GET ;
+        $view->preview($setting['displayId'], $setting['args']);
+        // reverting $_GET to backed state once the view is executed
+        $_GET = $_GETbkp;
+
+        // views which are empty will not have fieldsMarkup
+        $view->fieldsMarkup = isset($view->fieldsMarkup) ? $view->fieldsMarkup : [];
+        $currentViewData = [];
+        $fieldKeyForRows = isset($setting['fieldAsKeyForRows']) ? $setting['fieldAsKeyForRows'] : false;
+        // storing the markup for every row
+        foreach ($view->fieldsMarkup as $i => $row) {
+          // setting different key using the value of a field which identifies
+          // a row. Helpful when comparing view with multiple rows to handle
+          // deletions and insertions correctly
+          $index = $fieldKeyForRows ? $row[$fieldKeyForRows]->content : $i;
+          $currentViewData[$index] = [];
+          // storing the markup for every field
+          foreach ($row as $key => $fieldOutput) {
+            if ($key === $fieldKeyForRows) {
+              continue;
+            }
+            $fieldKey = $fieldOutput->label_html;
+            $currentViewData[$index][$fieldKey] = $fieldOutput->content;
+          }
+        }
+        $data[] = [
+          'view' => $setting['view'],
+          'data' => $currentViewData,
+          'title' => $view->get_title()
+        ];
+      }
+    }
+    return $data;
+  }
+
+  /**
+   * Stores data on property newData
+   *
+   * @return ViewsResultDiff
+   */
+  public function storeNewData() {
+    $this->newData = $this->getCurrentData();
```

Comment:
the whole distinction between `storeNewData` and `storeOldData` is confusing, this whole class is doing a lot of stuff, I'd suggest breaking it down into separate services like
```
ViewsDiffGenerator::getDiff($firstDataSet, $secondDataSet);
ViewsDataFetcher::getCurrentData($all, $params, $that, $are, $required);
```

Also the whole view setting structure is so complicated that it doesn't make any sense when I'm looking at it in the constructor. Maybe you could consider creating an object for this, in general this class is just hard to follow - it would be better to break it into more specific classes. I think diff generation is an ideal candidate for unit testing but I don't see any tests in this PR at all. With smaller, more specific classes we can write clear tests for them that don't require Drupal function calls.

Reply:
`ViewsDiffResultDiff` is a class that generates differences between views. I will create a class to fetch the data, and inject the results in `ViewsDiffResultDiff` to avoid external dependencies.